types/crossfilter2.d.ts

```diff
@@ -0,0 +1,103 @@
+declare namespace CrossFilter2 {
+
+    export interface Selector<T> {
+        (value: T): any;
+    }
+
+    export interface CrossFilterStatic {
+        <T>(data: T[]): CrossFilter<T>;
+        size(): number;
+        version: string;
+        permute<T>(arry: T[], index: number[], deep: number): T[];
+        heap: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Heap<T>;
+        }
+        heapselect: {
+            <T>(array: T[], lo: number, hi: number, k: number): T[];
+            by<T>(value: Selector<T>): HeapSelect<T>;
+        }
+        bisect: {
+            <T>(array: T[], value: T, lo: number, hi: number): number;
+            by<T>(value: Selector<T>): Bisector<T>;
+        }
+        insertionsort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(vaule: Selector<T>): Sort<T>;
+        }
+        quicksort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Sort<T>;
+        }
+    }
+
+    export interface Bisection<T> {
+        (array: T[], value: T, lo: number, hi: number): number;
+    }
+
+    export interface Bisector<T> extends Bisection<T> {
+        left: Bisection<T>
+        right: Bisection<T>
+    }
+
+    export interface Heap<T> {
+        (array: T[], lo: number, hi: number): T[];
+        sort(array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface HeapSelect<T> {
+        (array: T[], lo: number, hi: number, k: number): T[];
+    }
+
+    export interface Sort<T> {
+        (array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface CrossFilter<T> {
+        add(records: T[]): CrossFilter<T>;
+        remove(predicate?: Function): void;
```

Comment:
The function here is specifically a function from `T` to `boolean`. Can we express that with an interface?

Reply:
`remove()` takes a `Function` and `remove()` should return a `boolean`?  For the latter, I notice that Crossfilter fires a `triggerOnChange("dataRemoved")` event.  So should the `remove()` still return a `boolean`? For the former, from what I have found it seems that the code would have to be something like the following, but I'm not 100% sure this is a correct implementation?  Could someone verify, please?

```
export interface PredicateFn<T> {
        (data: T[], i: number) 
}
```

Reply:
No, I think `remove` does return void. The type of the predicate function is that it is a function from T -> boolean. I'm not sure if what you wrote here expresses that in Typescript or not, unfortunately.

Reply:
It could be expressed inline, like:
```
remove(predicate?: (data: T, i: number) => boolean): void;
```